clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1175,19 +1176,26 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
             try {
                 metadata.awaitUpdate(version, remainingWaitMs);
             } catch (TimeoutException ex) {
+                final String errorMessage = String.format("Topic %s not present in metadata after %d ms.",
+                        topic, maxWaitMs);
+                if (metadata.getError(topic) != null) {
+                    throw new TimeoutException(errorMessage, metadata.getError(topic).exception());
+                }
                 // Rethrow with original maxWaitMs to prevent logging exception with remainingWaitMs
```

Comment:
nit: this comment doesn't seem to apply here and should be above where the error message is defined.
Also just to be clear -- this code path is not one where we see anyting but retriable exceptions? I noticed the other path checks that but not this one.

Reply:
> Also just to be clear -- this code path is not one where we see anyting but retriable exceptions? I noticed the other path checks that but not this one.

Please see https://github.com/apache/kafka/pull/16344#discussion_r1648061530